### ISSUE TITLE
Idea: A dedicated adblock list

### ISSUE THREAD

**Issue by yokoffing** [hagezi/dns-blocklists#2346]
Hi. I was going to post this under discussions, but it looks like that tab isn't available.

### Dedicated Ad block List
Not that you don't already have a million things to do already... :upside_down_face:

In the spirit of efficiency: I was wondering if it would be worth it to make a dedicated ad block list, since most of the primary DNS lists are over 100k entries.

This new category of list would:
1. **Remove (not include) sources unrelated to web browser traffic**, such as device-specific trackers, browser and smart device telemetry, and IoT device tracking.
2. **Exclude commonly used lists** like EasyList and EasyPrivacy, if feasible.
3. **Be version-specific**: uBO version would omit uBO default lists, AdGuard's version would omit AdG's default lists, etc. 
4. **Focus on ad and tracking domains**, making it more suitable for ad blockers and complementary to existing lists.
   1. Remove (not include) TIF Medium.
   2. Remove (not include) security-related domains, leaving those for DNS-blocking --- or include them in a separate list for folks without DoH/DoT... But at that point, they could just use one of the five DNS lists.
      - e.g., I disable `Online Malicious URL Blocklist` in uBO and leave my DNS lists to filter malicious domains. They're hardly, if ever, triggered (thankfully).

For example, if we look at what's listed for the Pro++ list, here's sources we could cut out:

| **URL/File**                                           | **Remove?** | **Comments** |
|--------------------------------------------------------|-------------|--------------|
| Denylist Personal                                      |             |              |
| Extracted Wildcard Rules                             |             |   security related?            |
| Native Tracker Apple (iOS, macOS, tvOS)          |    x         |              |
| Native Tracker Amazon (Devices, Shopping, Video)  |   x          |              |
| Native Tracker Huawei (Devices)                  |         x    |              |
| Native Tracker LG webOS                         |    x         |              |
| Native Tracker TikTok (Fingerprinting)               |             |   check EasyPrivacy, or include just in case           |
| Native Tracker Microsoft (Windows, Office, MSN)      |     x        |              |
| Denylist Base                                        |             |              |
| Basic Extension                                      |             |              |
| Denylist Community non-Aggressive                    |             |              |
| Analytics/Metrics Extension                          |             |              |
| Balanced Extension                                   |             |              |
| Pop-Up Ads                                           |             |      |
| Advanced Extension                                   |             |              |
| AdGuard MultiParty                                   |             |      ...if not AdG          |
| EasyList MultiParty                                  |    x         |              |
| uBlock MultiParty                                    |            |    ... if not uBO          |
| Fake                                                 |             |      security related?             |
| Tracking Extension                                   |             |              |
| Aggressive Extension                                 |             |              |
| Native Tracker Vivo (EXPERIMENTAL)                   |     x        |              |
| Native Tracker OPPO/Realme (EXPERIMENTAL)            |      x       |              |
| Denylist Aggressive                                  |             |              |
| Denylist Community Aggressive                        |             |              |

It's not perfect. Might be easier to start with the Light list and slowly add/subtract from there. 

It would be your judgment call on the "extension" files, since I don't know what comprises them.

**Comment by hagezi** [comment]
whew, you're killing me ... :laughing:

**Comment by hagezi** [comment]
You want a Pro++ mini?

What list size are you aiming for? I have already optimised the size of the existing lists, malware is mainly only found in TIF medium and TIF. The problem is the popup ads list which is a mishmash of malicious and ads domains.

**Comment by hagezi** [comment]
Quick and Dirty check: `Pro++ - TIF - PopupAds - Gambling - AntiPiracy - Bugtracker - Telemetry - Native` = ~120000 domains

A lot of manual work will still have to be done. Probably too much has been removed and there are certainly still domains that can be removed.

The lists are DNS optimised, i.e. thousands of domains are saved by adding CNAMES. However, these are irrelevant for browser blockers.

I don't quite understand the point behind the idea, as the filter lists available for content blockers such as EasyList, AdGuard, uBlock and others are more than adequate. So when I use uBlock, almost nothing that comes from the browser is filtered by DNS. I only use my DNS lists in the DNS.

For DNS AdBlockers that can't process large lists, there are Light and Normal. Thanks to the optimisations, the Pro now has a size that means it also works in Little Snitch (<200000), for example. 

There is also Peter Lowe's Ad and tracking list if someone really only wants to block ads and tracking.

**Comment by hagezi** [comment]
I got as much out of the Light as I could, it couldn't be much smaller. It mainly only contains domains that have been on top lists (Umbrella, Cloudflare, Tranco, Majestic, ...) in the last 12 months. I have manually added domains that improve the compression so that the list is as small as possible. I could remove 7000 domains that are also on the TIF medium - but these domains are also on the top lists.

Native trackers are only partially included, bug trackers have been removed. The fake list is not included at all. At the moment I don't know where I should make savings without making it ineffective. With a lot of effort maybe 1000-2000 domains.

**Comment by hagezi** [comment]
[size optimization v1 light list](https://github.com/hagezi/dns-blocklists/commit/28c4b80c57b8b3c047661fd76227e0cfa4e76d50) 
new size: 58743 (-10000)

**Comment by hagezi** [comment]
new size: 57477 (removed malware, except badware)

Light now blocks Blocks Ads, Tracking, Metrics/Telemetry and some Badware.

**Comment by hagezi** [comment]
In the next step, I will run [AdGuard's Dead Domain Linter](https://github.com/AdguardTeam/DeadDomainsLinter) over the Light and see what else it does compared to my own deadcheck. I only remove domains with NXDOMAIN and SERVFAIL, the linter also removes domains that return NOERROR but have no A record. This is not really a problem with normal lists, but it is with malware lists.

**Comment by hagezi** [comment]
new size: 56683 (removed dead linter domains)

And now I can't think of any more ways to optimize the size without compromising the previous effectiveness.

**Comment by yokoffing** [comment]
> What list size are you aiming for?

Under 100k domains compressed.

> Peter Lowe's Ad and tracking

Default in uBO/Brave but not AdGuard. It's good for what it is, even if it only has 3k rules.

> I could use wildcard AdBlock rules in combination with allow rules, that would drastically reduce the size. However, the AdBlock version would then no longer be compatible with Pi-hole and the like, because the AdBlock DNS syntax is not fully implemented there.

> That's okay. This is a purely adblock list idea. The focus is: Here's an efficient ads and trackers list for your browser adblocker. Then you have your 5 DNS lists and TIF for the DNS-level (e.g., Pi-hole).

**Comment by hagezi** [comment]
@yokoffing https://raw.githubusercontent.com/hagezi/dns-blocklists/main/adblock/pro.plus.browser.txt

**Comment by yokoffing** [comment]
> https://raw.githubusercontent.com/hagezi/dns-blocklists/main/adblock/pro.plus.browser.txt

😮 I will give it a spin now!

**Comment by hagezi** [comment]
There is still some potential for optimisation in the list. It's a first throw ...

**Comment by yokoffing** [comment]
>still some potential for optimisation

You mean wildcard and allow rules?

### Quick Observations

- Is [Privacy Essentials](https://github.com/yokoffing/filterlists/blob/main/privacy_essentials.txt) still used as a source for Pro++? I noticed some first party rules not being blocked (e.g., `||segment.com^`). Of course, `$3p` rules are not included either, but I would expect that for a DNS-based list.
- Running the new Light Adblock list and Pro Plus Browser Edition list in uBO, the latter is detecting a lot more domains than the former, so that's to be expected.
- Pro Plus list has unique domains (compared to [all the lists I'm using](https://i.postimg.cc/9M3pcLkT/Screenshot-2024-03-16-at-11-46-28-u-Block-Dashboard.png)), so that's good.
- There's quite a bit of overlap with other lists (EasyPrivacy, AdGuard Tracking, Peter Lowe's), but it's not a pure one-to-one where if they have it then Pro Plus has it too. For instance, `||content22.online.citi.com^` is blocked by Peter Lowe + EasyPrivacy but not Pro Plus.
   - I think we'd have to be careful here. AdGuard has AdGuard Tracking Protection by default, but not EasyPrivacy. And that goes vice versa for uBO. So it'd be an error to exclude these list altogether and assume a user has enabled all of them.
      - **One solution:** You could make a `if` rule to remove domains if they appear on **both** EasyPrivacy and AdGuard Tracking. The same for EasyList and AdGuard Base filter.
      - **Another solution:** If you ever go into heavier optimization (uBO-specific list, AdG-specific list), that's where you could make some omissions based on their default lists.
     - You can have a default version that only assumes EasyList is enabled by default (AdBlock Plus?), because if many basic adblockers may only use one list, it's EasyList. And other adblockers, like built-in ones like Vivaldi, Opera, and Orion, support only the very basic of rules, so Pro Plus would benefit from more sources there.

**Comment by hagezi** [comment]
What I will definitely not do is remove rules that are on content filter lists that someone is theoretically already using. Doing this is a lot of work and as you do it, you do it wrong. Very few users will understand this and I can already see the issues because this and that is not blocked. Personally, I don't see the point either. For me, the effort and benefit are out of proportion.

I will offer a complete list that can be used in browser blockers in any combination. Optimised in size and optimised with AdBlock syntax.

What I have done so far:
- Used the compiled Pro++ as a base.
- Removed domains from the Pro++ that were not on my merged toplist in the last 12 months (~16 million domains).
- Removed TIF medium domains. 
- Added domains to improve the compression and thus the efficiency/size of the list.

Further potential for optimisation (later):
- Use of wildcard rules to further minimise the size and improving effectiveness.

From my point of view, there are now two effective lists available that can be used in ad blockers that cannot handle large lists or browser content blockers.

- Relaxed: Light
- Aggressive: Pro ++ mini (browser edition)

I will rename the Pro++ browser edition to Pro++ mini later, browser edition doesn't quite fit.

**Comment by hagezi** [comment]
https://raw.githubusercontent.com/hagezi/dns-blocklists/main/adblock/pro.plus.mini.txt

**Comment by cotusandro** [comment]
Please add a Wildcard Domains variant too.

**Comment by hagezi** [comment]
@cotusandro This is the plan, but the identification of the possible wildcard rules is manual work, anyone who wants to help is welcome.

**Comment by cotusandro** [comment]
The entries in the Adblock and Wildcard Domains variants are identical, except for the writing style. Since the content is the same, would it be feasible to automate the conversion of Adblock entries to the Wildcard Domains format? This could potentially simplify the process and reduce the manual work required for identifying wildcard rules.

**Comment by hagezi** [comment]
@cotusandro I think I misunderstood you, I meant wildcard adblock rules and reducing the size of the AdBlock format, e.g.

```
! Analytics
|analytics.*.*^
@@||analytics.controld.com^
@@||analytics.twitter.com^
|analytic.*.*^
|sanalytics.*.*^
|sanalytic.*.*^
|metrics.*.*^
|metric.*.*^
@@||metric.gstatic.com^
|smetrics.*.*^
|smetric.*.*^
|telemetry.*.*^
|clarity.*.*^
|matomo.*.*^
|plausible.*.*^
|posthog.*.*^
|piwik.*.*^
|utiq.*.*^
||*countly*.*^

! Bugtracking
|sentry.*.*^
||bugly.*.*^

! Amazon AWS
||*-adserver-*.amazonaws.com^
||*-analytics-*.amazonaws.com^
||*-beacon-*.amazonaws.com^
||*-collector-*.amazonaws.com^
||*-loggers-*.amazonaws.com^
||*-metrics-*.amazonaws.com^
||*-telemetry-*.amazonaws.com^
||adktracking-*.amazonaws.com^
||adn-tktracking-*.amazonaws.com^
||ads-*.amazonaws.com^
||adserver-*.amazonaws.com^
||ad-server-*.amazonaws.com^
||adservingvpc-*.amazonaws.com^
||adsk-*.amazonaws.com^
||beaconserver-*.amazonaws.com^
||log-collectors-*.amazonaws.com^
||popadex-*.amazonaws.com^
||rtbimp-loadbalancer-*.amazonaws.com^
||tracking-*.amazonaws.com^
||tracker-*.amazonaws.com^
||webanalytics-*.amazonaws.com^
||wildcard-ads-*.amazonaws.com^

! Google
||adservice.google.*^
||id.google.*^
```
The Pro++ mini is optimized for browsers specifically for adblockers that can process this format. The use of wildcard AdBlock rules makes the list incompatible for conversion to other formats.

What prevents you from using the DNS wildcard domain version of the normal Pro++?

**Comment by cotusandro** [comment]
I intended for this variant (https://raw.githubusercontent.com/hagezi/dns-blocklists/main/wildcard/pro.plus-onlydomains.txt). Pro++ Mini has the same blocking efficacy as the normal Pro++. I have noticed that a large number of domains included in Pro++ are parked domains, which would be the main cause of their 2-3x larger size compared to Mini. Again, since parked domains receive negligible traffic, they are not included in the top 1M list.

**Comment by hagezi** [comment]
I had already removed some, do you have a few more examples? How did you determine them?

It would also be interesting to determine these for the other lists.

**Comment by cotusandro** [comment]
I opened the first 20 entries and then a separate patch of 20 consecutive entries. I found that most of them are parked domains.

**Comment by cotusandro** [comment]
Most phishing, malware, scam, and fake domains are short-lived. Once these domains are noticed by blocklist maintainers, bad actors buy and move on to other domains. Although this increases the list size over time, it provides no value.
Most ads, affiliate, tracking, metrics, and telemetry domains are relatively long-lived and receive significant traffic. Many of them are included in the Top 1M lists.
I recommend using the Top 1M list for the lite, normal, pro, pro++, and ultimate versions, just as you have already done for the pro++ browser edition.

**Comment by hagezi** [comment]
@cotusandro I have added the other "wildcard" formats.
https://github.com/hagezi/dns-blocklists?tab=readme-ov-file#proplusmini

**Comment by jarelllama** [comment]
@hagezi @cotusandro , I experimented with parked domains detection for my Scam Blocklist. I also assumed many scam/malicious domains are parked and have been inflating my blocklist, so I came up with this simple detection code (I explain why I did not end up using it below):

``` bash
function check_for_parked {
    # Check for parked message in site's HTML
    while read -r domain; do
        if grep -qiFf "$parked_terms_file" <<< "$(curl -sL --max-time 1 "http://${domain}/")"; then
            printf "%s\n" "$domain" >> parked_domains.tmp  # Collate parked domains
        fi
    done < "$raw_file"
    [ ! -s parked_domains.tmp ]] && return  # Return if no parked domains found
    # Remove parked domains from raw file
    comm -23 "$raw_file" parked_domains.tmp > raw.tmp && mv raw.tmp "$raw_file"
    cat parked_domains.tmp >> "$parked_domains_file"  # Collate parked domains
}
```

It simply checks the site's HTML for common parked domain messages. I collated a few in the variable `$parked_terms_file`:
```
<!DOCTYPE html><html><head><script>window.onload=function(){window.location.href="/lander"}</script></head></html>
buy this domain
connectyourdomain
courtesy of godaddy.com
domain for sale
domain has expired
domain is for sale
domain is parked
domain may be for sale
domain parked
domain parking
error code: 1001
index of /
inquire about this domain
is currently parked
is this your domain
no files associated
page has been suspended
parked domain
parked-on-the-bun
parked_type
shop is currently unavailable
store under construction
temporarily parked
this account has been suspended
url=/cgi-sys/defaultwebpage.cgi
```

After testing with 500 domains, the false positive rate was 0% with about 1% false negatives that I couldn't detect without adding phrases that would have false positives. 

The problem lies in how long it takes to process each domain even with `curl --max-time 1`. If each domain took 1 second, my blocklist of 23000 would take 6 hours to process.

Out of the 500 domains I randomly pulled from my blocklist, less than 5% were parked. This small reduction in size did not seem worth the hours of running. So in the end I decided to not implement it. Regardless, I thought I would share the idea with you.

**Comment by hagezi** [comment]
Thanks @jarelllama, I'll have a look at that later. I had already thought about such a solution, but rejected it because of the running times.

**Comment by hagezi** [comment]
@jarelllama You may also find it interesting:

https://github.com/AdguardTeam/DeadDomainsLinter

**Comment by jarelllama** [comment]
> @jarelllama You may also find it interesting:
> 
> https://github.com/AdguardTeam/DeadDomainsLinter

I already discovered and implemented this a while back thanks to this thread :)

**Comment by hagezi** [comment]
Since the requests for the mini lists have increased since the Pro++ mini became available, I have decided to provide a mini version for the other versions as well:

For the Normal, the Light is the mini version. For the other lists (Pro, Pro++ and Ultimate), a mini version has also been made available.

https://github.com/hagezi/dns-blocklists#promini
https://github.com/hagezi/dns-blocklists#proplusmini
https://github.com/hagezi/dns-blocklists#ultimatemini
